aten/src/ATen/native/cuda/PersistentSoftmax.cuh

```diff
@@ -0,0 +1,390 @@
+
+template <typename input_t, typename output_t, typename acc_t, bool is_log_softmax>
+bool warp_softmax_backward_kernel(int log2_elements, int &warp_size, int &batches_per_warp, softmax_backward_func<input_t, output_t> &kernel) {
+    // determine size of a warp
+    const int next_power_of_two = 1 << log2_elements;
+    warp_size = (next_power_of_two < 32) ? next_power_of_two : 32;
+
+    // determine how many batches a warp should process.
+    batches_per_warp = (next_power_of_two <= 128) ? 2 : 1;
+
+    switch (log2_elements) {
+    case 0: // 1
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 2,1,1, is_log_softmax>;
+        break;
+    case 1: // 2
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 2,1,2, is_log_softmax>;
+        break;
+    case 2: // 4
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 2,1,4, is_log_softmax>;
+        break;
+    case 3: // 8
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 2,1,8, is_log_softmax>;
+        break;
+    case 4: // 16
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 2,1,16, is_log_softmax>;
+        break;
+    case 5: // 32
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 2,1,32, is_log_softmax>;
+        break;
+    case 6: // 64
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 2,2,32, is_log_softmax>;
+        break;
+    case 7: // 128
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 2,4,32, is_log_softmax>;
+        break;
+    case 8: // 256
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 1,8,32, is_log_softmax>;
+        break;
+    case 9: // 512
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 1,16,32, is_log_softmax>;
+        break;
+    case 10: // 1024
+        kernel = &softmax_warp_backward<input_t, output_t, acc_t, 1,32,32, is_log_softmax>;
+        break;
+    default:
+        return false;
+    }
+    return true;
+}
+
+template<typename input_t, typename output_t, typename acc_t, bool is_log_softmax>
+bool dispatch_softmax_backward(output_t *grad_input, const input_t *grad, const input_t *output, int softmax_elements, int softmax_elements_stride, int batch_count)
+{
+    if (softmax_elements == 0) {
+        return true;
+    } else if (softmax_elements <= 1024) {
+        // compute function index. there's a function for each power of two size up to 1024.
+        int log2_elements = 0;
+        while ((1 << log2_elements) < softmax_elements) ++log2_elements;
+
+        softmax_backward_func<input_t, output_t> kernel;
+        int warp_size, batches_per_warp;
+        if (!warp_softmax_backward_kernel<input_t, output_t, acc_t, is_log_softmax>(log2_elements, warp_size, batches_per_warp, kernel)) {
+            return false;
+        }
+
+        // use 128 threads per block to maximimize gpu utilization
+        constexpr int threads_per_block = 128;
```

Comment:
It looks like all these computationas are shared between forward/backward, so may be they can be abstracted away?

Reply:
Ok.